system/xmllib/libxml_xpathprocessor.cpp

```diff
@@ -775,6 +789,352 @@ extern ICompiledXpath* compileXpath(const char * xpath)
+}
+void addChildFromPtree(xmlNodePtr parent, IPropertyTree &tree, const char *name)
+{
+    if (isEmptyString(name))
+        name = tree.queryName();
+    if (*name==':') //invalid xml but maybe valid ptree?
+        name++;
+    StringAttr prefix;
+    const char *colon = strchr(name, ':');
+    if (colon)
+    {
+        prefix.set(name, colon-name);
+        name = colon+1;
+    }
+    xmlNodePtr node = xmlNewChild(parent, nullptr, (const xmlChar *)name, (const xmlChar *)tree.queryProp(nullptr));
+    if (!node)
+        return;
+    xmlNsPtr ns = nullptr;
+    if (tree.getAttributeCount())
+    {
+        Owned<IAttributeIterator> atts = tree.getAttributes();
+        ForEach(*atts)
+        {
+            const char *attname = atts->queryName()+1;
+            if (strncmp(attname, "xmlns", 5)==0)
+            {
+                attname+=5;
+                if (*attname==':')
+                    attname++;
+             xmlNewNs(node, (const xmlChar *)atts->queryValue(), (const xmlChar *)attname);
+            }
+            else
+                xmlNewProp(node, (const xmlChar *)attname, (const xmlChar *)atts->queryValue());
+        }
+    }
+    if (prefix.isEmpty())
+        node->ns = nullptr;
+    else
+    {
+        xmlNsPtr ns = xmlSearchNs(node->doc, node, (const xmlChar *)prefix.str());
+        if (!ns) //invalid but let's repair it
+            ns = xmlNewNs(node, (const xmlChar *)"urn:hpcc:unknown", (const xmlChar *)prefix.str());
+        node->ns = ns;
+    }
+    Owned<IPropertyTreeIterator> children = tree.getElements("*");
+    ForEach(*children)
+        addChildFromPtree(node, children->query(), nullptr);
+}
+
+IPropertyTree *createPtreeFromXmlNode(xmlNodePtr node);
+
+void copyXmlNode(IPropertyTree *tree, xmlNodePtr node)
+{
+    for(xmlAttrPtr att=node->properties;att!=nullptr; att=att->next)
+        tree->setProp((const char *)att->name, (const char *)xmlGetProp(node, att->name));
+
+    for (xmlNodePtr child=xmlFirstElementChild(node); child!=nullptr; child=xmlNextElementSibling(child))
+        tree->addPropTree((const char *)child->name, createPtreeFromXmlNode(child));
+}
+
+IPropertyTree *createPtreeFromXmlNode(xmlNodePtr node)
+{
+    if (!node)
+        return nullptr;
+    Owned<IPropertyTree> tree = createPTree((const char *)node->name);
+    copyXmlNode(tree, node);
+    return tree.getClear();
+}
+
+static const char *queryAttrValue(xmlNodePtr node, const char *name)
+{
+    if (!node)
+        return nullptr;
+    xmlAttrPtr att = xmlHasProp(node, (const xmlChar *)name);
+    if (!att)
+        return nullptr;
+    if (att->type != XML_ATTRIBUTE_NODE || att->children==nullptr || (att->children->type != XML_TEXT_NODE && att->children->type != XML_CDATA_SECTION_NODE))
+        return nullptr;
+    return (const char *) att->children->content;
+}
+
+class CEsdlScriptContext : public CInterfaceOf<IEsdlScriptContext>
+{
+private:
+    void *espCtx = nullptr;
```

Comment:
Why not store as an IEspContext *?

Reply:
Exactly I didn't want to expose the common library to ESP internals.  We could move things around or share some fairly isolated headers.  I think IEspContext may carry a lot of baggage.